src/messaging/ExchangeContext.cpp

```diff
@@ -139,12 +141,15 @@ CHIP_ERROR ExchangeContext::SendMessage(Protocols::Id protocolId, uint8_t msgTyp
 
     bool isStandaloneAck =
         (protocolId == Protocols::SecureChannel::Id) && msgType == to_underlying(Protocols::SecureChannel::MsgType::StandaloneAck);
+
+    //
+    // The caller clearly intends to send a message. Let's set the WillSendMessage flag so that if we fail at any
+    // point below, the exchange is still in a mode of being ready to send a message again, as well as being setup
+    // correctly for delegates to know that they still need to release the ref on this exchange.
+    //
     if (!isStandaloneAck)
     {
-        // If we were waiting for a message send, this is it.  Standalone acks
-        // are not application-level sends, which is why we don't allow those to
-        // clear the WillSendMessage flag.
-        mFlags.Clear(Flags::kFlagWillSendMessage);
+        mFlags.Set(Flags::kFlagWillSendMessage);
```

Comment:
Hmm.  So if I am responding to an incoming message and I try to send and fail, now it's my responsibility to explicitly close the exchange?  That's a change from the setup we had before (where the exchange would close unless there was a WillSend or a successful send), and I'm worried about it leaking for anyone who is not using SessionHolder.

Why do we need to do this, exactly?  So consumers don't have to think about whether they are responding to an incoming message or not when they get a failed send?

This needs an audit of existing SendMessage callers to make sure they are either using ExchangeHandle or following the new rules.  That is a must-fix.

Reply:
Wasn't the contract always that if `SendMessage` fails, that the caller has the responsibility of releasing the ref on the exchange? (i.e point # 2 in the rules section):

> If you send a message successfully that doesn't require a response, the ownership transfers to the ExchangeMgr, and it will close the exchange for you automatically.

Reply:
Will audit.

Reply:
> Wasn't the contract always that if SendMessage fails, that the caller has the responsibility of releasing the ref on the exchange? 

No, that was only if SendMessage fails when sending a message outside of incoming message processing.

Reply:
So, here's the coarse results of my audit of all callers to `SendMessage`:

1. `BdxOtaSender::HandleTransferSessionOutput`. This is not called from within `OnMessageReceived` processing, and is scheduled instead on a timer. The `TransferFacilitator` always calls `WillSendMessage` in its `OnMessageReceived` callback, further bolstering this. So this is not affected by this logic.
2. `BdxDownloader::HandleBdxEvent` calls `BDXDownloader::MessagingDelegate::SendMessage`, which the default impl of in `DefaultOTARequestor`, calls `ExchangeContext::SendMessage` and on error, will call `Close` on the EC. `HandleBdxEvent` is called from `PollTransferSession`, which is called from `OnMessageReceived`. Which means that this logic is actually already faulty relative to the existing assumptions...
3. `CASESession` and `PASESession` both adhere correctly to the existing assumptions. So that would need to get changed, assuming we follow through with the rules change.
4. IM logic use ExchangeHolders, so they're covered.

Reply:
Modulo the bugs that exist above, I'm fine removing this to return it back to the contract you've specified. That I think is the safest for now...

Reply:
Removed this bit of logic.

Reply:
> Which means that this logic is actually already faulty relative to the existing assumptions...

To be clear, calling Close there is safe, just not necessary.